project/ThreeMonth03/README.md

```diff
@@ -0,0 +1,121 @@
+# NSD Proposal
+
+## Basic Information
+
+I will implement conjugate gradient method as my final project.
+
+Project URL: [https://github.com/ThreeMonth03/Conjugate-Gradient-Method-library](https://github.com/ThreeMonth03/Conjugate-Gradient-Method-library)
+
+
+## Problem to Solve
+
+I want to solve the global minima of the quadratic function by searching the numerical solution. One of the algorithm that can solve the problem is conjugate gradient method.
+
+## Conjugate gradient method
+### Introduction
+
+Conjugate gradient method is a gradient-descent-based method that can find the minima of the quadratic function in the hyper-dimensional space. 
```

Comment:
What if the function is not quadratic?

Reply:
Conjugate gradient method cannot deal with non-quadratic function.

Reply:
Let me put it this way: What will happen when you use CG to find the minimum of a field that is not quadratic?

Reply:
After browsing the article on the net, 
I try to answer the reason why conjugate gradient method cannot solve non-quadratic function.

In fact, the original problem what CG method want to solve is $Ax = b$.
When $A$ is symmetric and positive-definite,
solving $Ax = b$ is equal to find the global minima of $f(x) = \frac{1}{2}x^{T}Hx + b^{T}x + c$, 
according to differentiating the quadratic function.

Thus, if the function is not quadratic, then the differential of the function is not $Ax = b$,
thus the CG method will be failed.

I hope the answer is helpful for you.

Reference: https://blog.csdn.net/Yemiekai/article/details/123999290

Reply:
After further reading,
I realize that I neglect the nonlinear conjugate gradient algorithm that solve non-linear target. 
Thus, if the input is non-quadratic funciton and there exists local minima in the function, 
it seems that nonlinear conjugate gradient algorithm can find the local minima.

Therefore, in order to avoid from the troublesome input format and gradient calculation, 
I decide to implement linear conjugate gradient algorithm, 
which can only find the global minimum of quadratic function according to the aforementioned reason.

The reference of non-linear conjugate function: https://indrag49.github.io/Numerical-Optimization/conjugate-gradient-methods-1.html#nonlinear-conjugate-gradient-algorithm

Reply:
Let me repeat my question because you did not answer it: What will happen when you use CG to find the minimum of a field that is not quadratic?

Reply:
Sorry for my poor English. I guess I answer the question this time. 

Before answering the question, I think I should provide additional information on CG method.
The CG method includes linear CG and non-linear CG algorithm, the primary difference of these methods are the calculations of the step lengths in every iteration.
Linear CG algorithm can precisely calculate the step length in every iteration, but in order to calculate the precise step length, the objective function could only be quadratic function.
Non-linear CG algorithm can approximately calculate the step length by line search, or gradient descent method, and the advantage of non-linear CG algorithm is that the target function could be convex nonlinear objective functions.

Therefore, if non-linear CG algorithm find the minimum $X$ of a field that is not quadratic, then the next iteration of the $X$ is also the minima because the gradient direction is zero vector. Everything is fine.
However, if we attempt to find the non-quadratic function by linear CG algorithm, because the linear CG algorithm cannot calculate the precise step length in every iteration, linear CG method will be failed. The reason why CG algorithm cannot calculate the precise step for non-quadratic function is that the optimization of the step length for non-quadratic function is not quite easy.

In conclusion, I think I should replace all the CG method with linear CG method in README.md, and modify the line 17 "Conjugate gradient method is a gradient-descent-based method" to "Linear conjugate gradient method is a  kind of steepest descent approach". I realize that I confuse gradient descent with steepest descent.

Thank you for your question.

Reply:
I think you are moving forward the idea in a right track.  As we discussed today, the key is that you need to consider the input data that are not a polynomial function.  How does your software systems handle it?

Reply:
If the input data that is not a polynomial function, the input should be a lot of the points of y and gradients. Thus, in theory, the non-linear conjugate gradient method can deal with the non-quadratic convex input function.

However, I think  some issues need considering when implementing the input data that is not a polynomial function. First, I think the user should input the data point and data gradient in every epoch by themselves, because the non-polynomial inputs, such as voxel and point cloud, are digital such that they are not precise enough. Additionally, the determination of the convex is a trouble.

In conclusion, I think I may deal with the non-quadratic convex input function to a certain degree. It may look like those in below. But I still cannot guarantee the probability of the convergence.

```c++
class CG_Algorithm{
     private:
           bool _is_poly_fucn;
           int epoch;
     public:
           double iteration(const vector<double> & point_xi, const vector<double> & gradient_xi){
                      //......
                      return point_x_i_plus_one;
           }
}
```

Reply:
A related question is in what format the input data should be.  After considering the question and summarizing what we have discussed, the proposal may be complete enough for development.

Keep in mind that a complete proposal (or plan) does not guarantee a successful project.  There are a lot of things to do in your execution.